File: backend/monitor.js
/**
 * StockPulse 24/7 Catalyst Monitor
 * Runs continuously, checks Finnhub for high-impact news, triggers Mac Desktop Notifications.
 */
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const configPath = path.join(__dirname, 'config.json');
const alertsPath = path.join(__dirname, 'alerts.json');

// ── Catalyst Engine Logic ──
// These are keywords that actually move markets significantly
const CATALYST_KEYWORDS = {
  "partnership": "High",
  "acquisition": "High",
  "merger": "High",
  "allocated": "High",
  "awarded": "High",
  "contract": "High",
  "earnings beat": "High",
  "fda approval": "High",
  "soars": "High",
  "plummets": "High",
  "guidance raised": "High",
  "ceo resigns": "High",
  "investigation": "High"
};

/**
 * Trigger a native macOS Desktop Notification
 */
function sendMacNotification(title, message) {
  // Escape quotes
  const safeTitle = title.replace(/"/g, '\\"');
  const safeMessage = message.replace(/"/g, '\\"');
  
  const script = `display notification "${safeMessage}" with title "StockPulse Alert: ${safeTitle}" sound name "Glass"`;
  
  exec(`osascript -e '${script}'`, (error) => {
    if (error) {
      console.error('Failed to send macOS notification:', error);
    }
  });
}

/**
 * Scan text for catalytic events
 */
function analyzeCatalyst(headline, summary) {
  const text = `${headline} ${summary}`.toLowerCase();
  
  let triggeredCatalyst = null;
  let impact = "Low";
  
  for (const [keyword, lvl] of Object.entries(CATALYST_KEYWORDS)) {
    if (text.includes(keyword)) {
      triggeredCatalyst = keyword;
      impact = lvl;
      break;
    }
  }
  
  if (triggeredCatalyst) {
    let prediction = `Positive momentum likely due to ${triggeredCatalyst}.`;
    if (['investigation', 'ceo resigns', 'plummets'].includes(triggeredCatalyst)) {
      prediction = `Negative momentum likely due to ${triggeredCatalyst}.`;
    }
    
    return {
      isCatalyst: true,
      keyword: triggeredCatalyst,
      impact: impact,
      prediction: prediction
    };
  }
  
  return { isCatalyst: false };
}

/**
 * Main Loop
 */
async function runMonitor() {
  console.log(`\n[${new Date().toLocaleTimeString()}] 🔍 Scanning for catalysts...`);
  
  // 1. Read Config
  let config;
  try {
    config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
  } catch (e) {
    console.error("❌ Could not read config.json. Please ensure it exists.");
    return;
  }

  if (!config.finnhub_api_key || config.finnhub_api_key === "YOUR_API_KEY_HERE") {
    console.error("❌ Missing API Key in backend/config.json");
    return;
  }

  // 2. Read previous alerts to prevent duplicates
  let previousAlerts = [];
  try {
    previousAlerts = JSON.parse(fs.readFileSync(alertsPath, 'utf8'));
  } catch (e) {
    previousAlerts = [];
  }

  const processedUrls = new Set(previousAlerts.map(a => a.url));
  let newAlertsFound = false;

  // 3. Scan Watchlist
  for (const ticker of config.watchlist) {
    const toObj = new Date();
    const fromObj = new Date();
    fromObj.setDate(fromObj.getDate() - 3); // Look at last 3 days of news for catalysts

    const to = toObj.toISOString().split('T')[0];
    const from = fromObj.toISOString().split('T')[0];

    const url = `https://finnhub.io/api/v1/company-news?symbol=${ticker}&from=${from}&to=${to}&token=${config.finnhub_api_key}`;
    
    try {
      const res = await fetch(url);
      const news = await res.json();
      
      if (!Array.isArray(news)) continue;

      // Only check recent 5 articles per scan to save processing
      const recentNews = news.slice(0, 5);

      for (const article of recentNews) {
        if (processedUrls.has(article.url)) continue; // Already alerted
        
        const analysis = analyzeCatalyst(article.headline, article.summary);
        
        if (analysis.isCatalyst) {
          console.log(`🚨 CATALYST FOUND FOR ${ticker}: ${analysis.keyword}`);
          
          const alertObj = {
            id: Date.now().toString(),
            ticker: ticker,
            timestamp: new Date().toISOString(),
            headline: article.headline,
            url: article.url,
            keyword: analysis.keyword,
            impact: analysis.impact,
            prediction: analysis.prediction
          };
          
          previousAlerts.unshift(alertObj); // Add to top
          processedUrls.add(article.url);
          newAlertsFound = true;
          
          // Trigger Mac Desktop Notification
          sendMacNotification(ticker, `${analysis.prediction}\n\n${article.headline}`);
        }
      }
    } catch (e) {
      console.error(`Error fetching news for ${ticker}:`, e.message);
    }
    
    // Tiny delay to respect API limits
    await new Promise(r => setTimeout(r, 200));
  }

  // 4. Save to JSON if new alerts found
  if (newAlertsFound) {
    // Keep only the last 100 alerts to prevent huge files
    const trimmedAlerts = previousAlerts.slice(0, 100);
    fs.writeFileSync(alertsPath, JSON.stringify(trimmedAlerts, null, 2));
    console.log(`✅ Saved new alerts to alerts.json`);
  } else {
    console.log(`💤 No new catalysts found.`);
  }
}

// Start the scheduler
console.log('==============================================');
console.log('🚀 StockPulse 24/7 Catalyst Monitor Started');
console.log('==============================================');
runMonitor();

// Schedule to run based on config
const intervalMinutes = 15; // default
setInterval(runMonitor, intervalMinutes * 60 * 1000);

File: backend/monitor.py
import os
import json
import time
import urllib.request
import urllib.parse
import xml.etree.ElementTree as ET
import subprocess
import sqlite3
from datetime import datetime, timedelta, timezone

# Paths
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
CONFIG_PATH = os.path.join(BASE_DIR, 'config.json')
DB_PATH = os.path.join(BASE_DIR, 'market_data.db')

def send_notification(title, text, ntfy_topic=None):
    # Mac local notification
    try:
        text_safe = text.replace('"', '\\"')
        title_safe = title.replace('"', '\\"')
        subprocess.run(['osascript', '-e', f'display notification "{text_safe}" with title "{title_safe}"'])
    except Exception:
        pass
        
    # NTFY Push notification (iOS)
    if ntfy_topic:
        try:
            req = urllib.request.Request(
                f"https://ntfy.sh/{urllib.parse.quote(ntfy_topic)}",
                data=text.encode('utf-8'),
                headers={
                    "Title": title.encode('utf-8'),
                    "Tags": "chart_with_upwards_trend,rotating_light"
                },
                method="POST"
            )
            urllib.request.urlopen(req, timeout=5)
        except Exception as e:
            print(f"Failed to send ntfy push: {e}")

def analyze_catalyst_with_gemini(headline, summary, gemini_key):
    """Scan text for catalytic events using Google Gemini 2.5 Flash"""
    if not gemini_key:
        return []
        
    url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={gemini_key}"
    
    prompt = (
        "You are a quantitative AI analyst. Read the following news headline and summary. "
        "Identify ALL publicly traded companies mentioned or implicitly affected. "
        "Identify ANY market-moving event, sentiment shift, or momentum driver (it does not need to be a massive catalyst, even minor news is okay). "
        "Return EXACTLY a JSON array of objects. Do not include markdown formatting. "
        "If no companies are affected, return []. "
        "Format for each object: "
        "{\"ticker\": \"<TICKER>\", \"isCatalyst\": true, \"keyword\": \"<1-3 word reason>\", \"impact\": \"<High/Medium/Low>\", \"prediction\": \"<1 sentence Bull/Bear thesis prediction on stock price momentum>\"} "
        f"Headline: {headline} | Summary: {summary}"
    )
    
    payload = {
        "contents": [{"parts": [{"text": prompt}]}]
    }
    
    import time
    try_count = 0
    while try_count < 3:
        try:
            req = urllib.request.Request(url, data=json.dumps(payload).encode('utf-8'), headers={'Content-Type': 'application/json'}, method='POST')
            with urllib.request.urlopen(req) as response:
                if response.getcode() == 200:
                    text_resp = response.read().decode()
                    data = json.loads(text_resp)
                    text_resp = data['candidates'][0]['content']['parts'][0]['text']
                    
                    clean_text = text_resp.replace('```json', '').replace('```', '').strip()
                    
                    if clean_text.startswith('{') and clean_text.endswith('}') and '},' in clean_text:
                        clean_text = f"[{clean_text}]"
                        
                    try:
                        result = json.loads(clean_text)
                    except Exception as parse_err:
                        return []
    
                    if isinstance(result, list):
                        return result
                    elif isinstance(result, dict) and result.get("isCatalyst"):
                        return [result]
                    else:
                        return []
                else:
                    return []
        except urllib.error.HTTPError as e:
            if e.code == 429:
                try_count += 1
                time.sleep(30) # Wait 30 seconds before retrying
            else:
                return []
        except Exception as e:
            return []
            
    return "RATE_LIMIT"

def run_monitor():
    current_time = datetime.now().strftime("%I:%M:%S %p")
    print(f"\n[{current_time}] 🔍 Scanning for catalysts...", flush=True)
    
    # 1. Read Config
    try:
        with open(CONFIG_PATH, 'r') as f:
            config = json.load(f)
    except Exception as e:
        print("❌ Could not read config.json. Please ensure it exists.")
        return

    if not config.get("gemini_api_key") or config["gemini_api_key"] == "YOUR_GEMINI_KEY":
        print("❌ Missing Gemini API Key in backend/config.json")
        return

    # 2. Read previous state from SQLite
    try:
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        
        # Make sure scanned_urls exists just in case
        cursor.execute('''CREATE TABLE IF NOT EXISTS scanned_urls (url TEXT PRIMARY KEY, timestamp TEXT NOT NULL)''')
        
        cursor.execute("SELECT url FROM scanned_urls")
        processed_urls = {row[0] for row in cursor.fetchall()}
    except Exception as e:
        print(f"Error reading from SQLite: {e}")
        processed_urls = set()
        
    new_alerts_found = False
    watchlist_updated = False

    # 3. Google News RSS Autonomous Search
    queries = [
        "tech stock momentum",
        "stock President OR partnership",
        "CEO endorsement OR acquisition rumor",
        "stock price target upgrade",
        "earnings beat stock surge",
        "new product launch tech stock"
    ]
    
    recent_news = []
    for query in queries:
        encoded_query = urllib.parse.quote_plus(query)
        url = f"https://news.google.com/rss/search?q={encoded_query}&hl=en-US&gl=US&ceid=US:en"
        try:
            req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
            with urllib.request.urlopen(req) as response:
                xml_data = response.read()
                root = ET.fromstring(xml_data)
                for item in root.findall('.//item')[:10]:
                    title = item.find('title').text
                    link = item.find('link').text
                    if link not in processed_urls:
                        recent_news.append({
                            "headline": title,
                            "summary": "",
                            "url": link
                        })
        except Exception as e:
            print(f"Error fetching Google News: {e}")

    for article in recent_news:
        article_url = article.get('url', '')
        headline = article.get('headline', '')
        summary = article.get('summary', '')
        
        analysis_results = analyze_catalyst_with_gemini(headline, summary, config.get("gemini_api_key"))
        
        if analysis_results == "RATE_LIMIT":
            print("⏳ Rate Limit Hit severely! Skipping remaining articles for this cycle...")
            break
            
        if not analysis_results:
            processed_urls.add(article_url)
            try:
                cursor.execute("INSERT OR IGNORE INTO scanned_urls (url, timestamp) VALUES (?, ?)", (article_url, datetime.utcnow().isoformat() + "Z"))
                conn.commit()
            except:
                pass
            time.sleep(1) # Tiny sleep just to be safe
            continue

        for analysis in analysis_results:
            if analysis and analysis.get('isCatalyst') and analysis.get('ticker'):
                ticker = analysis['ticker'].upper()
                keyword = analysis.get('keyword', 'UNKNOWN')
                print(f"🚨 AUTONOMOUS CATALYST FOUND FOR {ticker}: {keyword}", flush=True)
                
                # Auto-Track Watchlist Injection
                if ticker not in config.get('watchlist', []):
                    config.setdefault('watchlist', []).append(ticker)
                    watchlist_updated = True
                    print(f"➕ Auto-tracking new ticker: {ticker}")
                # 4. Fetch Exact Live Quote to record Alert Price
                alert_price = 0.0
                if config.get("finnhub_api_key"):
                    try:
                        quote_url = f"https://finnhub.io/api/v1/quote?symbol={ticker}&token={config['finnhub_api_key']}"
                        q_req = urllib.request.Request(quote_url, headers={'User-Agent': 'Mozilla/5.0'})
                        with urllib.request.urlopen(q_req) as q_resp:
                            q_data = json.loads(q_resp.read().decode())
                            alert_price = q_data.get('c', 0.0)
                    except Exception as q_err:
                        pass
                        
                # Insert into SQLite Alerts Table
                alert_id = str(int(time.time() * 1000))
                try:
                    cursor.execute('''
                    INSERT OR IGNORE INTO alerts 
                    (id, ticker, timestamp, date, headline, url, keyword, impact, prediction, alert_price)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    ''', (
                        alert_id,
                        ticker,
                        datetime.utcnow().isoformat() + "Z",
                        datetime.now().strftime('%Y-%m-%d'),
                        headline,
                        article_url,
                        keyword,
                        analysis.get('impact', 'High'),
                        analysis.get('prediction', ''),
                        alert_price
                    ))
                    
                    # Execute Mock Trade (Budget: $1000 per trade)
                    if alert_price and alert_price > 0:
                        shares = 1000.0 / alert_price
                        cursor.execute('''
                        INSERT INTO paper_trades 
                        (trade_id, ticker, action, price, shares, timestamp, status, alert_id)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                        ''', (
                            "TRD_" + alert_id,
                            ticker,
                            "BUY",
                            alert_price,
                            shares,
                            datetime.utcnow().isoformat() + "Z",
                            "OPEN",
                            alert_id
                        ))
                        
                    conn.commit()
                    new_alerts_found = True
                except Exception as e:
                    print(f"Failed to insert alert into DB: {e}")
                
                send_notification(ticker, f"MOCK BUY Executed: $1000 at ${alert_price} | {analysis.get('prediction', '')}\n\n{headline}", config.get("ntfy_topic"))
                
        processed_urls.add(article_url)
        try:
            cursor.execute("INSERT OR IGNORE INTO scanned_urls (url, timestamp) VALUES (?, ?)", (article_url, datetime.utcnow().isoformat() + "Z"))
            conn.commit()
        except:
            pass
            
        # Protect Gemini Free Tier Rate Limits (15 RPM -> 1 request every 4 seconds)
        time.sleep(5)
        
    if watchlist_updated:
        try:
            with open(CONFIG_PATH, 'w') as f:
                json.dump(config, f, indent=2)
        except Exception as e:
            print(f"Error updating config: {e}")
            
    try:
        conn.close()
    except:
        pass

    if new_alerts_found:
        print("✅ Saved new alerts to market_data.db", flush=True)
    else:
        print("💤 No new catalysts found.", flush=True)

if __name__ == "__main__":
    print("==============================================")
    print("🚀 MarketOracle 24/7 Catalyst Monitor Started")
    print("==============================================", flush=True)
    
    # Read config for interval
    try:
        with open(CONFIG_PATH, 'r') as f:
            cfg = json.load(f)
            interval_minutes = cfg.get("check_interval_minutes", 15)
    except:
        interval_minutes = 15
        
    # Initial run
    run_monitor()
    
    # Loop
    while True:
        time.sleep(interval_minutes * 60)
        run_monitor()
